### PR TITLE
feat: useKeyedState util

### PR DESCRIPTION
**Pull request by muselesscreator**
updates the useState mocking pattern to use a simple useKeyedState hook to lower overhead and allow final usage to follow rule-of-hooks.

Notes about the pattern update:
The new `useKeyedState` does not do anything with the passed key.  The key is just there so that when mocked, we can uniquely identify the different calls.
This is theoretically possible without keying them, but then the tests are dependent on the implementation detail of the order that the state values are initialized in, which absolutely should not matter to the tests.

This also allows us to have a single shared util to mock when setting up tests rather than having to do the self-import trick so that we can "test the exported version, which we can mock independently".  

This removes the need to test the definition/initialization of separate state hooks, as we only ever use `useKeyedState`, which is a shallow wrapper around `React.useState`.

Finally, when actually using the hooks, devs will need to still use a `use<Hook>` hook, which benefits from the react checking for how/when hooks should be used.

### CODE REVIEW

File: src/examples/hooks.js [hunk 1]
```diff
@@ -2,24 +2,22 @@
...
+});
 
 export const formUrl = 'localhost:18000/form-url';
 
 export const useExampleComponentData = () => {
-  const [importClicked, setImportClicked] = module.state.importedClicked(0);
-  const [fileChanged, setFileChanged] = module.state.fileInputChanged(null);
-  const [, setLoaded] = module.state.loaded(false);
-  const [, setNumEvents] = module.state.numEvents(0);
+  const [importClicked, setImportClicked] = useKeyedState(stateKeys.importedClicked, 0);
```
Comment (leangseu-edx) [inline src/examples/hooks.js:17]: This feel unnecessary for testing. The stateKeys provide no value to the actual code. I think we should just test with values.

```js
state.hasBeenIntializeWith([
  0,
  null,
  false,
  0,
]);
```

Comment (muselesscreator) [reply src/examples/hooks.js:17]: but the order should not matter.

Comment (muselesscreator) [reply src/examples/hooks.js:17]: we want to check on the connected value by function/purpose, not by the order it was initialized

Comment (muselesscreator) [reply src/examples/hooks.js:17]: tests should not fail for re-ordering of the state initializations, which have no effect on behavior or output.

**Review by leangseu-edx — APPROVED** [review]
LGTM 👍

**Pull request merged**